update cloud controller manager docs for v1.8

andrewsykim opened the pull request:
Initial documentation for cloud controller manager in v1.8. There are definitely a lot of improvements to make, but wanted to the first draft in sooner rather than later. As @luxas mentioned previously, I think it would be super helpful to have some guides for setting up CCM (using kubeadm, etc), I decided to leave that for another PR and link to it from this doc.

cc @luxas @wlan0 @jhorwit2 @prydie @liggitt

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/5400)
<!-- Reviewable:end -->

tengqm commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
+
+Keep in mind that setting up your cluster to use cloud controller manager will change your cluster behaviour in a few ways:
+
+* kubelets specifying `--cloud-provider=external` will add a [taint during initialization](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/well_known_labels.go#L51-L55). This marks the node as needing a second initialization from an external controller before it can be scheduled work. Note that in the event that cloud controller manager is not available, new no ...
```
Maybe we should avoid referring to the source code directly since not all the readers are developers. An alternative would be explicitly spell out the taint: `node.cloudprovider.kubernetes.io/uninitialized: true` with an effect `NoSchedule`.

tengqm commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
+* node controller - responsible for updating nodes using cloud APIs.
+* service controller - responsible for loadbalancers on your cloud against services of type LoadBalancer.
+* route controller - responsible for setting up network routes on your cloud
+* persistent volume label controller - responsible for labeling persistent volumes on your cloud - ensure that the persistent volume label admission plugin is not enabled on your kube-apiserver.
```
Better spell out the name of the admission controller plugin: `PersistentVolumeLabel` and add a link, i.e.
```
[PersistentVolumeLabel](/docs/admin/admission-controllers/#persistentvolumelabel) 
```

luxas replied:
I realized that we have to remove this from the kubeadm admission chain as well... For v1.9

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
+{% capture overview %}
+**Cloud Controller Manager is an alpha feature in 1.8. In upcoming releases it will
+be the preferred way to integrate Kubernetes with any cloud. This will ensure cloud providers
+can develop their features independantly from [Kubernetes core](https://github.com/kubernetes/kubernetes).
```
end the ** here, otherwise the whole doc gets bold

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
-The methods in [cloudprovider.Interface](https://git.k8s.io/kubernetes/pkg/cloudprovider/cloud.go) are self-explanatory. All of the
-[existing providers](https://git.k8s.io/kubernetes/pkg/cloudprovider/providers) satisfy this interface. If your cloud is already a part
-of the existing providers, you do not need to write a new provider; you can proceed directly with linking your cloud provider to the `cloud-controller-manager`.
+Every cloud has their own set of requirements for running `cloud-controller-manager`, it should not be too different from the requirements when running `kube-controller-manager`. As a general rule of thumb you'll need:
```
their own cloud provider integration (instead of `cloud-controller-manager`)

luxas replied:
Describe this as it should behave as a normal Deployment that can be `kubectl apply`-d to the cluster

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
 
-Once your code is ready, you must import that code into `cloud-controller-manager`. See the [rancher cloud sample](https://github.com/rancher/rancher-cloud-controller-manager) for a reference example. The import step in the sample is the only step required to link your cloud provider to the `cloud-controller-manager`.
+* cloud authentication/authorization: your cloud may require a token or IAM rules to allow access to their APIs
+* kubernetes authentication/authorization: cloud-controller-manager may need abac/rbac set to speak to the kubernetes apiserver
```
abac is deprecated.
Mention "RBAC rules" only

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
-To run `cloud-controller-manager`, add it to your existing Kubernetes cluster as a Master component. All other master components except `kube-controller-manager` can be run without any changes.
+Successfully running cloud-controller-manager requires some changes to your cluster configuration.
+
+* `kube-apiserver` and `kube-controller-manager` must either specify no cloud provider or the `external` cloud provider for the flag `--cloud-provider`. This ensures that it does not run any cloud specific loops that would be run by cloud controller manager.
```
must not specify a `--cloud-provider` flag. In the future, this flag will be deprecated and removed.

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
+* persistent volume label controller - responsible for labeling persistent volumes on your cloud - ensure that the persistent volume label admission plugin is not enabled on your kube-apiserver.
+
+
+## Examples
```
Can we make a generic variant of this somehow?
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: cloud-controller-manager
  namespace: kube-system
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1beta1
metadata:
  name: system:cloud-controller-manager
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: cloud-controller-manager
  namespace: kube-system
---
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    app: cloud-controller-manager
  name: cloud-controller-manager
  namespace: kube-system
spec:
  selector:
    matchLabels:
      app: cloud-controller-manager
  template:
    metadata:
      labels:
        app: cloud-controller-manager
    spec:
      serviceAccountName: cloud-controller-manager
      containers:
      - name: cloud-controller-manager
        image: gcr.io/google_containers/cloud-controller-manager:v1.7.0
        command:
        - /usr/local/bin/cloud-controller-manager
        - --cloud-provider=gce
        - --allocate-node-cidrs=true
        - --configure-cloud-routes=true
        - --cluster-cidr=10.135.0.0/16
        volumeMounts:
        - name: etcssl
          mountPath: /etc/ssl
      tolerations:
      - key: node.cloudprovider.kubernetes.io/uninitialized
        value: "true"
        effect: NoSchedule
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
      volumes:
      - name: etcssl
        hostPath:
          path: /etc/ssl
```

In order to provide an example. For instance, parameterize out `--cloud-provider` etc

andrewsykim replied:
Let's add an example file somewhere in kubernetes/kubernetes and just link to it

andrewsykim replied:
@luxas added examples here https://github.com/kubernetes/kubernetes/pull/52374 and I'll link them once they're in master

jhorwit2 replied:
What's the best place to provide comments on this example? I'll put it here but can move it if some where else is better. 

* Is `cluster-admin` still the preferred way to handle the controller manager (and in this case the ccm) as well or should we be using a service account per controller? cc @liggitt 
* The `app` label should be `k8s-app`.
* Why is the example a DS instead of a deployment like the self-hosted controller manager? We can use anti-affinity to make sure they are spread. 
* The `etcssl` volume should be read only.

liggitt replied:
* I'd expect the cloud controller manager to have a defined role, not use cluster-admin
* The CCM also supports (and should use) `--use-service-account-credentials` to run the individual control loops with specific service accounts

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
+* service controller - responsible for loadbalancers on your cloud against services of type LoadBalancer.
+* route controller - responsible for setting up network routes on your cloud
+* persistent volume label controller - responsible for labeling persistent volumes on your cloud - ensure that the persistent volume label admission plugin is not enabled on your kube-apiserver.
+
```
And anything else you want to do! You're out of core :smile:

andrewsykim replied:
I'll leave this as a side note :P

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
+
+### Background
+
+Before going into how to build your own cloud controller manager, some background on how it works under the hood is helpful. The cloud controller manager is code from `kube-controller-manager` utilizing Go interfaces to allow implementations from any cloud to be plugged in. Most of the logic for the cloud controller manager will still live in Kubernetes core but the implementations for each cloud ...
```
Most of the _scaffolding_ and generic controller implementation will be in core; in the cloud controller _library_.
But it will always "exec out" to the interface you give to it

_OR_ you can dismiss that and write your ccm from scratch, your choice

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,34 +1,85 @@
 ---
 approvers:
```
You can add myself and @wlan0 here as well

andrewsykim replied:
done

luxas commented:
Nit: I think this should be retargeted to the release-1.8 branch. Is that still true @chenopis ?

wlan0 commented:
I think this should be re-targeted to release 1.8 branch as well. 

@andrewsykim Thanks for updating this. I think it looks good overall :+1:  

I would add one more section about its limitations. 

One important limitations is that cloud volumes will not be usable with the CCM. For eg. GCE PD/AWS EBS volumes will not be usable if you're running CCM today. 

This is because setting `--cloud-provider=external` in the KCM disables the volume controller. The volume controller is responsible for provisioning, and attaching/detaching volumes. But all is not lost - we have a plan for transitioning the users to using the CCM without losing functionality. @cheftako is working on a new flag called `--external-plugin` for the KCM to allow the KCM volume controller to load EBS/PD logic without pulling the entire cloudprovider. I believe that this would be ready by 1.9. @cheftako can confirm.

This is an interim measure for existing cloud specific volumes providers. For all others volume types, and for newer clouds, users may use flex volumes.

andrewsykim commented:
I was considering adding a section for limitations but figured it may be diving too deep into internals. Looking back it seems mentioning the future plan for volumes is pretty important as it would block a lot of people from using cloud controller manager going forward. 

Some other limitations I think we should add:
* scalability - for larger clusters, instead of using bandwidth from kubelets via local metadata, we now have ccm making those calls on their behalf which could end up being a cluster bottle neck. 
* chicken and egg problems - there are several chicken and egg problems with this architecture. A prominent one being we can't do TLS bootstrapping on kubelet yet cause the kubelet doesn't know it's address types. 
Of course we're doing work to address these limitations, but it might be worth mentioning so people are aware. 
 
What does everyone think?

jhorwit2 commented:
> scalability - for larger clusters, instead of using bandwidth from kubelets via local metadata, we now have ccm making those calls on their behalf which could end up being a cluster bottle neck.

not if they use `--provider-id` and set that based on local metadata

> chicken and egg problems - there are several chicken and egg problems with this architecture. A prominent one being we can't do TLS bootstrapping on kubelet yet cause the kubelet doesn't know it's address types.

I think that's very important to call out. I was looking at using that feature until I realized there is an open issue with the CCM.

jhorwit2 commented:
I think we should add some notes about `--provider-id` and `--node-ip` flags on the kubelet. I don't see those anywhere.

andrewsykim commented:
@jhorwit2 I'm a bit conflicted with documenting `--provider-id`. It definitely addresses the scalability concerns but it significantly increases operational overhead required to setup CCM.  Also there seems to be a chance that the provider-id format may change, in which case we're better off not advertising it til then. This is just my personal take on it though, I'll add it if we all agree it's worth mentioning :) @wlan0 @luxas

jhorwit2 commented:
> there seems to be a chance that the provider-id format may change

Is that ticketed some where? I don't remember any discussion around that. That'd be a breaking change to existing providers so I'm curious about that.

andrewsykim commented:
No issue as far as I know, I'm mostly speculating ;). I know auto-detect is deprecated now, but there are currently two different formats in the code base that I know of ([here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2097-L2103) and [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go#L73-L84)). Seems like the one in `pkg/cloudprovider` is authoritative but I'm curious to see what other formats are being used right now or in the past.

wlan0 commented:
> I was considering adding a section for limitations but figured it may be diving too deep into internals. Looking back it seems mentioning the future plan for volumes is pretty important as it would block a lot of people from using cloud controller manager going forward.

I agree. I think it's important for people to know what they are getting into with the CCM. It is better to let them know upfront, than them realizing that it doesn't support a particular capability, and then being surprised and frustrated by this effort. 

> scalability - for larger clusters, instead of using bandwidth from kubelets via local metadata, we now have ccm making those calls on their behalf which could end up being a cluster bottle neck.
chicken and egg problems - there are several chicken and egg problems with this architecture. A prominent one being we can't do TLS bootstrapping on kubelet yet cause the kubelet doesn't know it's address types.

These are important operational concerns to be aware of. 

> Of course we're doing work to address these limitations, but it might be worth mentioning so people are aware.

Adding to your point, we are addressing volumes. With @cheftako's work on `--external-plugin` flag for KCM (coming soon), users won't lose this functionality as soon as that work is merged.

andrewsykim commented:
Thanks for the feedback everyone, I'll make updates tomorrow after the meeting.

luxas commented:
@andrewsykim Any update?

chenopis commented:
I have moved this PR to the `release-1.8` branch.

andrewsykim commented:
Sorry folks, had a busy week! I'll address comments this weekend! :)

jhorwit2 commented on docs/tasks/administer-cluster/developing-cloud-controller-manager.md:
```diff
@@ -0,0 +1,43 @@
+2. Use [main.go in cloud-controller-manager](https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/controller-manager.go) from Kubernestes core as a template for your main.go. As mentioned above, the only difference should be the cloud package that will be imported.
+3. Import your cloud package in `main.go`, ensure your package has an `init` block to run [cloudprovider.RegisterCloudProvider](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/plugins.go#L42-L52).
+
+Use [DigitalOcean](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/master/main.go) and [Racher](https://github.com/rancher/rancher-cloud-controller-manager/blob/master/main.go) cloud controller managers as examples!
```
We should add more examples. 

* [rancher](https://github.com/rancher/rancher-cloud-controller-manager/)
* [keepalived](https://github.com/munnerz/keepalived-cloud-provider)

andrewsykim replied:
rancher is in there already, should probably use point form to make it clearer, didn't know about keepalived.

jhorwit2 commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+Kubernetes v1.6 contains a new binary called `cloud-controller-manager`. `cloud-controller-manager` is a daemon that embeds cloud-specific control loops. These cloud-specific control loops were originally in the `kube-controller-manager`. Since cloud providers develop and release at a different pace compared to the Kubernetes project, abstracting the provider-specific code to the `cloud-controlle ...
+
+The `cloud-controller-manager` can be linked to any cloud provider that satisifies [cloudprovider.Interface](https://git.k8s.io/kubernetes/pkg/cloudprovider/cloud.go). For backwards compatibility, the [cloud-controller-manager](https://github.com/kubernetes/kubernetes/tree/master/cmd/cloud-controller-manager) provided in the core Kubernetes project uses the same cloud libraries as `kube-controlle ...
```
nit: s/in-project/in tree to keep the terminology the same

jhorwit2 commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
-
-Once your code is ready, you must import that code into `cloud-controller-manager`. See the [rancher cloud sample](https://github.com/rancher/rancher-cloud-controller-manager) for a reference example. The import step in the sample is the only step required to link your cloud provider to the `cloud-controller-manager`.
+* cloud authentication/authorization: your cloud may require a token or IAM rules to allow access to their APIs
+* kubernetes authentication/authorization: cloud-controller-manager may need RBAC rules set to speak to the kubernetes apiserver
```
nit: RBAC isn't the only way to do authorization

andrewsykim replied:
https://github.com/kubernetes/kubernetes.github.io/pull/5400#discussion_r138439033

jhorwit2 replied:
I had no idea it was deprecated. Thanks :)

andrewsykim replied:
Same, I think at it's an unofficial deprecation at the moment?

jhorwit2 commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
-Once your code is ready, you must import that code into `cloud-controller-manager`. See the [rancher cloud sample](https://github.com/rancher/rancher-cloud-controller-manager) for a reference example. The import step in the sample is the only step required to link your cloud provider to the `cloud-controller-manager`.
+* cloud authentication/authorization: your cloud may require a token or IAM rules to allow access to their APIs
+* kubernetes authentication/authorization: cloud-controller-manager may need RBAC rules set to speak to the kubernetes apiserver
+* high availabilty: like kube-controller-manager, you may want a high available setup for cloud controller manager using leader election.
```
Should mention the `cloud-controller-manager` uses leader election by default and the name of the configmap (and flag if there is one to disable it)

andrewsykim replied:
Does it use leader election by default?

jhorwit2 replied:
[It appears to](https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/options/options.go#L64)

jhorwit2 commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
-To run `cloud-controller-manager`, add it to your existing Kubernetes cluster as a Master component. All other master components except `kube-controller-manager` can be run without any changes.
+Successfully running cloud-controller-manager requires some changes to your cluster configuration.
+
+* `kube-apiserver` and `kube-controller-manager` must not specify a `--cloud-provider` flag or must set it to `--cloud-provider=external`. This ensures that it does not run any cloud specific loops that would be run by cloud controller manager. In the future, this flag will be deprecated and removed.
```
Should we recommend to always explicitly set the flag to external rather than optionally setting it and just leaving it blank.

luxas replied:
s/must not specify a `--cloud-provider` flag or must set it to `--cloud-provider=external`/MUST NOT specify the `--cloud-provider` flag"/

andrewsykim commented:
Comments addressed, separated docs for administering CCM and developing CCM. Added examples too, PTAL.

jhorwit2 commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+As of v1.8, cloud controller manager can implement:
+
+* node controller - responsible for updating nodes using cloud APIs.
```
It can also delete nodes if they no longer have a good state in the cloud API.

andrewsykim replied:
update to delete :P

jhorwit2 commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+### Scalability
+
+In the previous architecture for cloud providers, we relied on kubelets using a local metadata service to retreive node information about itself. With this new architecture, we now fully rely on the cloud controller managers to retrieve information for all nodes. For very larger clusters, you should consider possible bottle necks such as resource requirements and API rate limiting.
```
the `--provider-id` flag alleviates this concern. Should we call this out?

andrewsykim replied:
I think we should have better supporting docs for what the format for --provider-id should be first, what do you think?

jhorwit2 replied:
Sounds good to me.

jhorwit2 commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+The goal of the cloud controller manager project is to decouple development of cloud features from the core Kubernetes project. Unforunately, many aspects of the Kubernetes project has assumptions that cloud provider features are tightly integrated into the project. As a result, adopting this new architecture can create several situations where a requests is being made for information from a clou ...
+
+A good example of this is the TLS bootstrapping feature in the Kubelet. Currently, TLS bootstraping assumes that the Kubelet has the ability to ask the cloud provider (or a local metadata service) for all its address types (private, public, etc) but cloud controller manager cannot set a node's address types without being initialzed in the first place which requires that the kubelet has TLS certif ...
```
Is there a ticket tracking the TLS bootstrap issue? All i can find in the [beta ticket](https://github.com/kubernetes/kubernetes/issues/48690) is a link to a comment. It'd be nice to link here to provide more info if the user wants it as well as a place to subscribe for updates.

andrewsykim replied:
cc @liggitt

luxas replied:
Don't have that yet I think.
AI for the WG I guess?

jhorwit2 commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+kind: DaemonSet
+metadata:
+  labels:
+    app: cloud-controller-manager
```
We should use [`k8s-app`](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/kube-proxy/kube-proxy-ds.yaml#L8).

jhorwit2 commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+      serviceAccountName: cloud-controller-manager
+      containers:
+      - name: cloud-controller-manager
+        image: gcr.io/google_containers/cloud-controller-manager:v1.7.0
```
Should we use `<YOUR_CLOUD_PROVIDER_IMAGE>` or something like below with the `--cloud-provider` flag.

andrewsykim replied:
no because this is the in-tree example

jhorwit2 commented on docs/tasks/administer-cluster/cloud-controller-manager-deployment-example.yaml:
```diff
@@ -0,0 +1,59 @@
+        app: cloud-controller-manager
+    spec:
+      serviceAccountName: cloud-controller-manager
+      containers:
```
Should this also use the node selector to force it to run on the masters and the taint?

andrewsykim replied:
I don't think so, if you wanted that you should use daemonset

jhorwit2 replied:
From what I've seen it's fairly common to limit control plane components to only master nodes regardless if they are a daemonset or deployment. [Bootkube](https://github.com/kubernetes-incubator/bootkube/blob/4dd7a4a575d0a0a4f24f2452180c4416c52f842a/pkg/asset/internal/templates.go#L419) does it and I _think_ kubeadm does as well.

andrewsykim replied:
Good point, though I think it serves as a good example then to leave this as a Deployment on non-master nodes since not everyone has their master nodes as schueable. You can also use the Daemonset example to set the node roles and tolerations yourself since these examples aren't really meant to work out of the box anyways, more as guidelines.

luxas commented on docs/tasks/administer-cluster/cloud-controller-manager-deployment-example.yaml:
```diff
@@ -0,0 +1,59 @@
+# This is an example of how to setup cloud-controller-manager as a Deployment in your cluster.
```
I'm not sure we need two examples... it's straightforward to do the conversion from DS to Deployment IAC, but high maintenance cost

luxas replied:
ping, can you remove this?

luxas commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+      serviceAccountName: cloud-controller-manager
+      containers:
+      - name: cloud-controller-manager
+        image: gcr.io/google_containers/cloud-controller-manager:v1.7.0
```
v1.8.0?

luxas commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+        image: gcr.io/google_containers/cloud-controller-manager:v1.7.0
+        command:
+        - /usr/local/bin/cloud-controller-manager
+        - --cloud-provider=<YOUR_CLOUD_PROVIDER>   # Add your own cloud provider here!
```
Add note that if you use `gcr.io/google_containers/cloud-controller-manager:v1.x.0`, it may be an in-tree one, if you use your own image with your own provider, it can be your own ;)

luxas commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+      - key: node-role.kubernetes.io/master
+        effect: NoSchedule
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
```
Add a comment that this makes this DS land only on masters

luxas commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+      - key: node.cloudprovider.kubernetes.io/uninitialized
+        value: "true"
+        effect: NoSchedule
+      - key: node-role.kubernetes.io/master
```
Add note that this makes it possible to schedule on the master

luxas commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+        - --configure-cloud-routes=true
+        - --cluster-cidr=172.17.0.0/16
+      tolerations:
+      - key: node.cloudprovider.kubernetes.io/uninitialized
```
Add note that this makes it possible for the ccm to bootstrap itself or sth like that

luxas commented on docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml:
```diff
@@ -0,0 +1,63 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: cluster-admin
```
Add BIG WARNING that this is DISCOURAGED or just link to a `system:cloud-controller-manager` ClusterRole and bind to that, and say to the user: fill in necessary verbs/resources here

andrewsykim replied:
I think I'm going to add an actual ClusterRole in this example, but I need to hash out what verbs/resources it actually needs access to. I'm going to do some testing and find out, mind if I updated this in another PR?

liggitt replied:
what API calls are cloud controllers responsible for making?

andrewsykim replied:
IIRC:
* list/watch/update/delete nodes 
* list/watch/update services
* list/watch/update PVL

^ there could be more, I want to test it to be sure

jhorwit2 replied:
I haven't tested it but I think this is what the CCM needs currently. (was planning to next week)
```
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - get
  - list
  - watch
  - update
  - patch
  - delete
- apiGroups:
  - ""
  resources:
  - services
  verbs:
  - list
  - watch
  - patch
- apiGroups:
  - ""
  resources:
  - persistentvolumes
  verbs:
  - list
  - watch
  - patch
```

andrewsykim replied:
Nice! This will be a good starting point for my tests too, thanks @jhorwit2!

jhorwit2 replied:
We could also use @liggitt's awesome new project to generate it :) https://github.com/liggitt/audit2rbac

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+Successfully running cloud-controller-manager requires some changes to your cluster configuration.
+
+* `kube-apiserver` and `kube-controller-manager` must not specify a `--cloud-provider` flag or must set it to `--cloud-provider=external`. This ensures that it does not run any cloud specific loops that would be run by cloud controller manager. In the future, this flag will be deprecated and removed.
+* `kubelet` must run with `--cloud-provider=external`. This is to ensure that the kubelet is aware that it must be initialized by the cloud controller manager before it is scheduled any work.
```
MAY BE run with `--cloud-provider=external` if your cloud needs to initialize nodes before becoming schedulable.
If that's not required, don't set the flag

andrewsykim replied:
I would consider `--cloud-provider=external` is a requirement  since `AddCloudNode` is a no-op otherwise https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L280-L284 and nodecontroller is a very large part of CCM

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+* kubelets specifying `--cloud-provider=external` will add a taint `node.cloudprovider.kubernetes.io/uninitialized` with an effect `NoSchedule` during initialization. This marks the node as needing a second initialization from an external controller before it can be scheduled work. Note that in the event that cloud controller manager is not available, new nodes in the cluster will be left unsched ...
+* cloud information about nodes in the cluster will no longer be retrieved using local metadata, but instead all API calls to retreive node information will go through cloud controller manager. This may mean you can restrict access to your cloud API on the kubelets for better security. For larger clusters you may want to consider if cloud controller manager will hit rate limits since it is now re ...
+* kube-apiserver should not run the Persistent Volume Label admission controller either since the cloud-controller-manager takes over labeling persistent volumes. To prevent the Persistent Volume Label admission plugin from running, make sure the kube-apiserver has a --admission-control flag with a value that does not include PersistentVolumeLabel.
```
move this to the section above and say "... SHOULD NOT `PersistentVolumeLabel` admission controller since..."

andrewsykim replied:
ahh, good catch, put this in the wrong section

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+* service controller - responsible for loadbalancers on your cloud against services of type LoadBalancer.
+* route controller - responsible for setting up network routes on your cloud
+* [PersistentVolumeLabel Admission Controller](/docs/admin/admission-controllers/#persistentvolumelabel) - responsible for labeling persistent volumes on your cloud - ensure that the persistent volume label admission plugin is not enabled on your kube-apiserver.
+
```
- your_new_custom_feature_here_if_out_of_tree

luxas commented on docs/tasks/administer-cluster/developing-cloud-controller-manager.md:
```diff
@@ -0,0 +1,46 @@
+
+### In Tree
+
+For in-tree cloud providers, you can run the in-tree cloud controller manager as a [Deployment](/docs/tasks/administer-cluster/cloud-controller-manager-deployment-example.yaml) or [Daemonset](/docs/tasks/administer-cluster/cloud-controller-manager-daemonset-example.yaml) in your cluster.
```
Just reference the running-the-ccm doc

rrati commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+* kubelets specifying `--cloud-provider=external` will add a taint `node.cloudprovider.kubernetes.io/uninitialized` with an effect `NoSchedule` during initialization. This marks the node as needing a second initialization from an external controller before it can be scheduled work. Note that in the event that cloud controller manager is not available, new nodes in the cluster will be left unsched ...
+* cloud information about nodes in the cluster will no longer be retrieved using local metadata, but instead all API calls to retreive node information will go through cloud controller manager. This may mean you can restrict access to your cloud API on the kubelets for better security. For larger clusters you may want to consider if cloud controller manager will hit rate limits since it is now re ...
+* kube-apiserver should not run the Persistent Volume Label admission controller either since the cloud-controller-manager takes over labeling persistent volumes. To prevent the Persistent Volume Label admission plugin from running, make sure the kube-apiserver has a --admission-control flag with a value that does not include PersistentVolumeLabel.
```
You also need to create an InitializerConfiguration so that the Persistent Volume Label controller will actually act on new persistent volumes.  See https://github.com/kubernetes/kubernetes.github.io/blob/release-1.8/docs/tasks/administer-cluster/running-cloud-controller.md

rrati commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+## Limitations
+
+Running cloud controller manager comes with a possible limitations. Although these limitations are being in addressed in upcoming releases, it's important that you are aware of these limitations for production workloads.
```
with a few possible limitations ?

rrati replied:
are being addressed in upcoming releases (remove extra "in")

rrati commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+### Support for Volumes
+
+Cloud controller manager does not implement any of the volume controllers found in `kube-controller-manager` as volume integrations also require coordination with kubelets. As we evolve CSI (container storage interface) and add stronger support for flex volume plugins, necessary support will be added to cloud controller manager so that clouds can fully integrate with volumes.
```
Can we add a link to the CSI docs/proposal?

rrati commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,96 @@
+
+### Chicken and Egg
+
+The goal of the cloud controller manager project is to decouple development of cloud features from the core Kubernetes project. Unforunately, many aspects of the Kubernetes project has assumptions that cloud provider features are tightly integrated into the project. As a result, adopting this new architecture can create several situations where a requests is being made for information from a clou ...
```
where a request is being made ...

steveperry-53 commented:
@andrewsykim Please rebase your branch so that it picks up 99fbc2b. That should fix the deploy/netlify failure. Thanks.

andrewsykim commented:
@steveperry-53 can we cherry pick that into the `release-1.8` branch please?

steveperry-53 commented:
@andrewsykim I don't understand your question about cherry picking 99fbc2b into the release-1.8 branch. That commit is already in the release-1.8 branch.

andrewsykim commented:
@steveperry-53 yeah my bad, I didn't realize we updated release-1.8 with master. Thanks!

luxas commented:
@andrewsykim Please ping me when this can be re-reviewed

jhorwit2 commented:
I feel like we should add another page like the [kubelet](https://kubernetes.io/docs/admin/kubelet/) with a brief synopsis and the config options.

luxas commented:
@jhorwit2 I think that's auto-generated by cobra or something like that.
Can you look into it?

jhorwit2 commented:
@luxas I reached out in #sig-docs on slack. @steveperry-53 is looking into it for all the reference docs. @wlan0 added the support to generate the markdown docs back in June.

andrewsykim commented:
@luxas @liggit @rrati @jhorwit2 addressed comments, PTAL

Here are the previews for the docs:
https://deploy-preview-5400--kubernetes-io-vnext-staging.netlify.com/docs/tasks/administer-cluster/running-cloud-controller/
https://deploy-preview-5400--kubernetes-io-vnext-staging.netlify.com/docs/tasks/administer-cluster/developing-cloud-controller-manager/

jhorwit2 commented on docs/tasks/administer-cluster/developing-cloud-controller-manager.md:
```diff
@@ -0,0 +1,40 @@
+* [DigitalOcean](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/master/main.go)
+* [keepalived](https://github.com/munnerz/keepalived-cloud-provider/blob/master/main.go)
+* [Rancher](https://github.com/rancher/rancher-cloud-controller-manager/blob/master/main.go)
+
```
We should link to one page or the other to provide examples so they don't get out of sync (the other page is missing keepalived).

rrati commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,90 @@
+* `kube-apiserver` and `kube-controller-manager` MUST NOT specify the --cloud-provider flag. This ensures that it does not run any cloud specific loops that would be run by cloud controller manager. In the future, this flag will be deprecated and removed.
+* `kubelet` must run with `--cloud-provider=external`. This is to ensure that the kubelet is aware that it must be initialized by the cloud controller manager before it is scheduled any work.
+* kube-apiserver SHOULD NOT run the Persistent Volume Label admission controller since the cloud-controller-manager takes over labeling persistent volumes. To prevent the Persistent Volume Label admission plugin from running, make sure the kube-apiserver has a --admission-control flag with a value that does not include PersistentVolumeLabel.
+* For the `cloud-controller-manager` to label persistent volumes, initializers will need to be enabled and an InitializerConifguration needs to be added to the system.  Follow [these instructions](/docs/admin/extensible-admission-controllers.md#enable-initializers-alpha-feature) to enable initializers.  Use the following YAML to create the InitializerConfiguration:
```
The link to how to enable initializers doesn't seem to be hoping to the specific section of the linked document.  It goes to the correct page, but would be nice to hop to the specific section on how to enable initializers.  I could've sworn it did this before

rrati commented:
@andrewsykim lgtm

luxas commented on docs/tasks/administer-cluster/running-cloud-controller.md:
```diff
@@ -1,38 +1,90 @@
+
+For cloud controller managers not in Kubernetes core, you can find the respective projects in repos maintained by cloud vendors or sig leads.
+
+* [DigitalOcean](https://github.com/digitalocean/digitalocean-cloud-controller-manager)
```
as @jhorwit2 said,  you have two identical lists here, should be one...

andrewsykim replied:
fixed

andrewsykim commented:
Comments addressed PTAL
